llama_index/query_engine/retry_query_engine.py

```diff
@@ -58,3 +61,63 @@ def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
     async def _aquery(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
         """Not supported."""
         return self._query(query_bundle)
+
+
+class RetryGuidelineQueryEngine(BaseQueryEngine):
+    """Does retry with evaluator feedback
+    if query engine fails evaluation.
+
+    Args:
+        query_engine (BaseQueryEngine): A query engine object
+        guideline_evaluator (GuidelineEvaluator): A guideline evaluator object
+        resynthesize_query (bool): Whether to resynthesize query
+        max_retries (int): Maximum number of retries
+        callback_manager (Optional[CallbackManager]): A callback manager object
+    """
+
+    def __init__(
+        self,
+        query_engine: BaseQueryEngine,
+        guideline_evaluator: GuidelineEvaluator,
+        resynthesize_query: bool = False,
+        max_retries: int = 3,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        self._query_engine = query_engine
+        self._guideline_evaluator = guideline_evaluator
+        self.max_retries = max_retries
+        self.resynthesize_query = resynthesize_query
+        super().__init__(callback_manager)
+
+    def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Answer a query."""
+        response = self._query_engine._query(query_bundle)
+        if self.max_retries <= 0:
+            return response
+        typed_response = (
+            response if isinstance(response, Response) else response.get_response()
+        )
+        query_str = query_bundle.query_str
+        eval = self._guideline_evaluator.evaluate_response(query_str, typed_response)
+        if eval.passing:
+            logger.debug("Evaluation returned True.")
+            return response
+        else:
+            logger.debug("Evaluation returned False.")
+            new_query_engine = RetryGuidelineQueryEngine(
+                self._query_engine,
+                self._guideline_evaluator,
+                self.resynthesize_query,
+                self.max_retries - 1,
+                self.callback_manager,
+            )
+            query_transformer = FeedbackQueryTransformation(
```

Comment:
make this an optional param that can be specified in the `__init__`. 

can default to this if user doesn't specify. allows thme to customize the query transformation they'd want

Reply:
Hmm so the problem is currently the FeedbackQueryTransformation constructor requires an evaluation object (!= evaluator). I could move the evaluator into the object so it just passes on a good evaluator.

Reply:
Hmm currently constructor for FeedbackQueryTransformation takes an evaluation which is only possible after querying. Let me see if I can abstract it out.